### PR TITLE
Updates Kubelet Plugin Registration process

### PR DESCRIPTION
**Pull request by claudiubelu**
#### What type of PR is this?

/kind failing-test
/kind bug

/sig windows
/sig testing

/priority important-soon
/milestone v1.27

#### What this PR does / why we need it:

Currently, when a Kubelet Plugin is being added in the DesiredStateOfWorld, a timestamp is saved in the PluginInfo. This timestamp is then updated on subsequent plugin reregistrations.

The Reconciler, when it detects different timestamps for a Plugin in its DesiredStateOfWorld and ActualStateOfWorld, it will then trigger a Plugin unregister and then a new Plugin registration.

Basically, the timestamp is being used to detect whether or not a Plugin needs to be reregistered or not. However, this can be an issue on Windows, where the time measurements are not as fine-grained. time.Now() calls within the same ~1-15ms window will have the same timestamp. This can mean that Plugin Reregistration events can be missed on Windows [1]. Because of this, some of the Plugin registration unit tests fail on Windows.

This commit updates the behaviour, instead of relying on different timestamps, the Reconciler will check the set PluginInfo UUID to detect a Plugin Reregistration. With this change, the unit tests mentioned above will also pass on Windows.

[1] https://github.com/golang/go/issues/8687

#### Which issue(s) this PR fixes:

Related: https://github.com/kubernetes/kubernetes/issues/51540

#### Does this PR introduce a user-facing change?

```release-note
Kubelet Plugins are now re-registered properly  on Windows if the re-registration period is < 15ms.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by claudiubelu** [comment]
@SergeyKanzhelev, have you had a change to review this?

**Comment by claudiubelu** [comment]
@SergeyKanzhelev, did you have a chance to look over this?

**Comment by claudiubelu** [comment]
Note: The Windows unit tests ran on this PR [1], and the following tests that are enabled in this PR passed:

```
k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache: Test_ASW_PluginExistsWithCorrectUUID_Negative_WrongUUID
k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache: Test_DSW_AddOrUpdatePlugin_Positive_ExistingPlugin
k8s.io/kubernetes/pkg/kubelet/pluginmanager/reconciler: Test_Run_Positive_Register
k8s.io/kubernetes/pkg/kubelet/pluginmanager/reconciler: Test_Run_Positive_RegisterThenUnregister
k8s.io/kubernetes/pkg/kubelet/pluginmanager/reconciler: Test_Run_Positive_ReRegister
```

[1] https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114136/pull-ci-kubernetes-unit-windows/1618920350374957056

File: pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go
Comment (bart0sh) [inline pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go:104]: @claudiubelu Can you explain why do we want to keep timestamp and pass it around if we switch to using UUID? Do we want to keep using timestamp on non-windows platforms?

File: pkg/kubelet/pluginmanager/cache/actual_state_of_world_test.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/pluginmanager/cache/actual_state_of_world_test.go:171]: "newer UUID" sounds wrong. Maybe different?

File: pkg/kubelet/pluginmanager/cache/actual_state_of_world.go
Comment (claudiubelu) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world.go]: Done

File: pkg/kubelet/pluginmanager/cache/actual_state_of_world_test.go
Comment (claudiubelu) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world_test.go]: Done

File: pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go
Comment (claudiubelu) [reply pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go]: Sure. The ``oe.operationGenerator.GenerateRegisterPluginFunc`` will call ``actualStateOfWorldUpdater.AddPlugin``, with a ``PluginInfo`` that will have a Timestamp set [1]. The ``PluginInfo``'s timestamp could still be useful information.

[1] https://github.com/kubernetes/kubernetes/blob/f97d14c6c88e92cb505f8a9147294705a93247e3/pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go#L115

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

File: pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go
Comment (bart0sh) [reply pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go:104]: > The PluginInfo's timestamp could still be useful information.

How it could be useful if it's not used anywhere? I'd suggest to remove it to make an API cleaner.

Comment (claudiubelu) [reply pkg/kubelet/pluginmanager/operationexecutor/operation_executor.go]: Done

File: pkg/kubelet/pluginmanager/cache/actual_state_of_world.go
Comment (bart0sh) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:57]: Why do we want to deprecate internal APIs? We can just remove it as it's not used anywhere. Am I missing something?

**Comment by bart0sh** [comment]
@claudiubelu please, address this comment, thanks: https://github.com/kubernetes/kubernetes/pull/114136#discussion_r1101237824

File: pkg/kubelet/pluginmanager/cache/actual_state_of_world.go
Comment (claudiubelu) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:57]: This was suggested during the SIG node meting on January 31st (https://docs.google.com/document/d/1Ne57gvidMEWXR70OxxnRkYquAoMpt56o75oZtg-OeBg/edit#):

```
[claudiubelu] Proposed changes to how kubelet detects updates for registered plugins b/c current implementation doesn’t work on Windows due to timestamp granularity issues https://github.com/kubernetes/kubernetes/pull/114136 
https://cs.k8s.io/?q=PluginExistsWithCorrectTimestamp&i=nope&files=&excludeFiles=&repos=
https://testgrid.k8s.io/sig-windows-signal#windows-unit-master test failure caused by timestamp
this has been proposed and brought up in the past, but no reviews came in.
it is one of the last consistently failing tests in the testgrid above.
[MikeBrown] consider keeping existing public method and add another one, deprecate first one later. nod just to protect custom plugin managers test buckets etc .. deprecate first thx.. see:
[https://cs.k8s.io/?q=PluginExistsWithCorrectTimestamp&i=nope&files=&excludeFiles=&repos=](https://cs.k8s.io/?q&#61;PluginExistsWithCorrectTimestamp&i&#61;nope&files&#61;&excludeFiles&#61;&repos&#61;)
```

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

File: pkg/kubelet/pluginmanager/cache/actual_state_of_world.go
Comment (bart0sh) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:57]: @claudiubelu ok, thanks. So, this API is used somewhere else, it seems. Can you point me out to where it's used?

Comment (bart0sh) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:57]: @mikebrow can you help? ^^^^

Comment (SergeyKanzhelev) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:57]: I cannot find good examples either. I'd be ok to keep it as deprecated for now just in case and maybe remove in couple of versions.

Comment (bart0sh) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:57]: I'd suggest to either remove it or deprecate **and** create an issue. Otherwise it has big chances to be forgotten.

Comment (claudiubelu) [reply pkg/kubelet/pluginmanager/cache/actual_state_of_world.go:57]: > I'd suggest to either remove it or deprecate **and** create an issue. Otherwise it has big chances to be forgotten.

Created issue: https://github.com/kubernetes/kubernetes/issues/115897

**Comment by valaparthvi** [comment]
Hi @claudiubelu. With Code Freeze coming in on March 15, please ensure the PR goes in before that if you would like for it to be a part of v1.27 release. Post Code Freeze, we will only be able to accept Release Blockers and Critical issues.

**Comment by claudiubelu** [comment]
> Hi @claudiubelu. With Code Freeze coming in on March 15, please ensure the PR goes in before that if you would like for it to be a part of v1.27 release. Post Code Freeze, we will only be able to accept Release Blockers and Critical issues.

The PR received some reviews from sig-node, waiting approvals.

**Comment by liggitt** [comment]
/milestone v1.27

actually, I can't tell if this is still targeting 1.27... it seems like a big change to merge this late, but I can't tell how bad the bug is

**Comment by liggitt** [comment]
/milestone clear

since this has gone a few weeks without activity or indication it will be approved for 1.27, and does not look like it is addressing an issue introduced in 1.27

**Comment by SergeyKanzhelev** [comment]
> /milestone v1.28

Is it MUST have for 1.28? We typically use milestone closer to the end of a release to track required work

**Comment by claudiubelu** [comment]
@SergeyKanzhelev, anything blocking this PR?

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Comment by SergeyKanzhelev** [comment]
> @SergeyKanzhelev, anything blocking this PR?

we need approver here. The change and the follow up bug created lgtm

**Comment by claudiubelu** [comment]
Can we get an approver for this?

**Comment by sftim** [comment]
Would this PR need a changelog (could an end user detect the changed behavior)?

**Comment by jsturtevant** [comment]
/assign @mrunalp @saad-ali 

@claudiubelu code you add a release note if needed?

**Comment by zelenushechka** [comment]
Hello!
Release Signal shadow here.
This issue has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting [02:00 UTC Wednesday 6th March 2024 / 18:00 PDT Tuesday 5th March 2024](https://everytimezone.com/s/5056bace) (less than a month from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As this issue is tagged for 1.30, is it still planned for this release?

**Comment by claudiubelu** [comment]
> Hello! Release Signal shadow here. This issue has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting [02:00 UTC Wednesday 6th March 2024 / 18:00 PDT Tuesday 5th March 2024](https://everytimezone.com/s/5056bace) (less than a month from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
> 
> As this issue is tagged for 1.30, is it still planned for this release?

Yes. It is planned for quite a few releases already. Waiting for it to be merged.

**Comment by zelenushechka** [comment]
Hello!
Release Signal shadow here.
I'd like to check what's the status. The code freeze is starting [02:00 UTC Wednesday 6th March 2024 / 18:00 PDT Tuesday 5th March 2024](https://everytimezone.com/s/5056bace) in a week, and we want to ensure that each PR has a chance to be merged on time.

**Comment by pacoxu** [comment]
clear the milestone as it has no update for week.
/milestone v1.31
Hope we can make it in next release cycle.
As this is a bugfix, we may backport it later.

**Comment by Vyom-Yadav** [comment]
Hello @claudiubelu ! This PR has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting `02:00 UTC Wednesday 10th July 2024` (about `4 weeks` from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for `1.31`, is it still planned for this release?

**Comment by claudiubelu** [comment]
> Hello @claudiubelu ! This PR has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting `02:00 UTC Wednesday 10th July 2024` (about `4 weeks` from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
> 
> As the PR is tagged for `1.31`, is it still planned for this release?

Hasn't been updated because I got no reviews. Waiting for those.

FWIW, it seems the unit tests are passing on Windows as well.

**Comment by Vyom-Yadav** [comment]
Hello @claudiubelu,

This PR hasn't been updated in a while. What's the current status? Do you want us to reach out to the folks who are supposed to review this? 

Reminder: Code freeze starts `02:00 UTC Wednesday 24th July 2024 / 19:00 PDT Tuesday 23rd July 2024` (about 1 week from now). Please make sure the PR has both `lgtm` and `approved` labels before the code freeze.

**Comment by Vyom-Yadav** [comment]
/milestone clear
We are way past the code freeze. Please add the milestone for the next release if you'd like to target that.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
